notify-debouncer-refined/src/lib.rs

```diff
@@ -0,0 +1,788 @@
+                _ => true,
+            } {
+                queue.events.push_back((ts, event));
+            }
+        } else {
+            self.queues.insert(
+                path.to_path_buf(),
+                Queue {
+                    events: [(ts, event)].into(),
+                },
+            );
+        }
+    }
+}
+
+/// Debouncer guard, stops the debouncer on drop.
+pub struct Debouncer<T: Watcher, C: FileIdCache> {
+    watcher: T,
+    debouncer_thread: Option<std::thread::JoinHandle<()>>,
+    data: DebounceData<C>,
+    stop: Arc<AtomicBool>,
+}
+
+impl<T: Watcher, C: FileIdCache> Debouncer<T, C> {
+    /// Stop the debouncer, waits for the event thread to finish.
+    /// May block for the duration of one tick_rate.
+    pub fn stop(mut self) {
+        self.set_stop();
+        if let Some(t) = self.debouncer_thread.take() {
+            let _ = t.join();
+        }
+    }
+
+    /// Stop the debouncer, does not wait for the event thread to finish.
+    pub fn stop_nonblocking(self) {
+        self.set_stop();
+    }
+
+    fn set_stop(&self) {
+        self.stop.store(true, Ordering::Relaxed);
+    }
+
+    /// Access to the internally used notify Watcher backend
+    pub fn watcher(&mut self) -> &mut T {
+        &mut self.watcher
+    }
+
+    /// Access to the internally used notify Watcher backend
+    pub fn cache(&mut self) -> MappedMutexGuard<C> {
+        MutexGuard::map(self.data.lock(), |data| &mut data.cache)
+    }
+}
+
+impl<T: Watcher, C: FileIdCache> Drop for Debouncer<T, C> {
+    fn drop(&mut self) {
+        self.set_stop();
+    }
+}
+
+/// Creates a new debounced watcher with custom configuration.
+///
+/// Timeout is the amount of time after which a debounced event is emitted.
+///
+/// If tick_rate is None, notify will select a tick rate that is 1/4 of the provided timeout.
+pub fn new_debouncer_opt<F: DebounceEventHandler, T: Watcher, C: FileIdCache + Send + 'static>(
+    timeout: Duration,
+    tick_rate: Option<Duration>,
+    mut event_handler: F,
+    file_id_cache: C,
+    config: notify::Config,
+) -> Result<Debouncer<T, C>, Error> {
+    let data = Arc::new(Mutex::new(DebounceDataInner::new(file_id_cache, timeout)));
+    let stop = Arc::new(AtomicBool::new(false));
+
+    let tick_div = 4;
+    let tick = match tick_rate {
+        Some(v) => {
+            if v > timeout {
+                return Err(Error::new(ErrorKind::Generic(format!(
+                    "Invalid tick_rate, tick rate {:?} > {:?} timeout!",
+                    v, timeout
+                ))));
+            }
+            v
+        }
+        None => timeout.checked_div(tick_div).ok_or_else(|| {
+            Error::new(ErrorKind::Generic(format!(
+                "Failed to calculate tick as {:?}/{}!",
+                timeout, tick_div
+            )))
+        })?,
+    };
+
+    let data_c = data.clone();
+    let stop_c = stop.clone();
+    let thread = std::thread::Builder::new()
```

Comment:
if you want to, you can already move ahead of #467 and use a tick-less system - otherwise we could introduce this later on

the idea is to either let a new event wake up the debouncer, or schedule a sleep-timeout if we need another tick (remaining events in the event queue)

I'll merge #467 after this one for the mini debouncer, and can also do the followup for this one, if you want

Reply:
I only saw #467 after I finished my PR. I think we should merge it first and then improve it.